File: typescript/examples/basic.ts
/**
 * basic.ts - download one thumbnail to a file.
 *
 * Usage:
 *   npx tsx basic.ts https://www.python.org/static/img/python-logo.png thumb.jpg
 */

import { writeFileSync } from "node:fs";
import { Client } from "../src/index.js";

async function thumbnail(url: string, path: string): Promise<void> {
  const tbr = await new Client().verify();
  const result = await tbr.thumb(url);

  const m = result.media;
  writeFileSync(path, m?.thumbnail.bytes ?? new Uint8Array(0));
  console.log(
    `${m?.kind ?? "?"}  ` +
      `${m?.fileSize?.toLocaleString() ?? "?"} bytes  ->  ` +
      `${m?.thumbnail.length.toLocaleString() ?? "0"} bytes  ` +
      `(${result.source ?? "render"})  ` +
      `${path}`,
  );
}

async function main(): Promise<void> {
  const [url, path] = process.argv.slice(2);
  if (!url || !path) {
    console.error("usage: npx tsx basic.ts <url> <out.jpg>");
    process.exit(2);
  }
  await thumbnail(url, path);
}

main().catch((err) => {
  console.error(err.message);
  process.exit(1);
});

File: rust/examples/basic.rs
//! basic.rs — download one thumbnail to a file.
//!
//! Usage:
//!     cargo run --example basic https://www.python.org/static/img/python-logo.png thumb.jpg

use std::env;
use std::fs;
use thumbrella::Client;

#[tokio::main]
async fn main() {
    let args: Vec<String> = env::args().collect();
    if args.len() != 3 {
        eprintln!("usage: {} <url> <out.jpg>", args[0]);
        std::process::exit(2);
    }
    let url = &args[1];
    let path = &args[2];

    if let Err(err) = thumbnail(url, path).await {
        eprintln!("{err}");
        std::process::exit(1);
    }
}

async fn thumbnail(url: &str, path: &str) -> Result<(), Box<dyn std::error::Error>> {
    // Client reads TBR_CONNECT env var for server URL or cloud token.
    // verify() ensures the connection is good before proceeding.
    let tbr = Client::new(None);
    tbr.verify().await?;

    // thumb() auto-verifies — returns Err on failure instead of a
    // placeholder like batch() or stream() would.
    let result = tbr.thumb(url).await?;
    let guard = result.lock().unwrap();

    fs::write(path, guard.thumbnail.bytes())?;
    println!(
        "{}  {:>8}  ->  {:>5} bytes  ({})  {path}",
        guard.kind.as_deref().unwrap_or("?"),
        guard.file_size.map_or("?".into(), |s| s.to_string()),
        guard.thumbnail.len(),
        guard.source.as_deref().unwrap_or("render"),
    );

    Ok(())
}
